### PR TITLE
Add `status` to the variations REST API

### LINKED ISSUE #15215: Variations REST API 'Visible' Prop Incorrect
There are a couple issues with the [`visible` prop on the variations endpoint](https://woocommerce.github.io/woocommerce-rest-api-docs/#product-variation-properties):

1) The description reads `Define if the attribute is visible on the "Additional information" tab in the product's page`. This is what the `visible` prop on an attribute does, not a variation.

2) What actually gets returned in the response is [catalog visibility](https://github.com/woocommerce/woocommerce/blob/f4048380602bd41137d9d3a0cd04981778327373/includes/abstracts/abstract-wc-product.php#L1383). Variation's don't have a catalog visibility.

3) When updating the `visible` prop using the variations endpoint, [it actually updates the post status](https://github.com/woocommerce/woocommerce/blob/master/includes/api/class-wc-rest-product-variations-controller.php#L251). There is no way to get the post status of a variation via the API right now (which I need and will be opening a proper PR up for).

We need to remove the visible prop in a future API version.

### PR DESCRIPTION
**Pull request by emdashcodes**
This adds a `status` field to a variation's response, which represents post status (published, private, etc). WooCommerce uses 'status' to enable or disable variations without deleting them (see the `enabled` checkbox in `wp-admin`). We need to be able to properly view and update the status in Calypso. This also adds the ability to update the status without relying on the confusing 'visible' prop, which this PR removes (fixes #15215).

Since the variations endpoint calls `parent::prepare_objects_query` we don't need to do anything extra to query for a certain status. This is already supported by the endpoint.

To Test:
* Run `phpunit`.
* Make sure all tests pass.

### CODE REVIEW

File: includes/api/class-wc-rest-product-variations-controller.php [hunk 1]
```diff
@@ -248,11 +249,17 @@ protected function prepare_object_for_database( $request, $creating = false ) {
 
 		$variation->set_parent_id( absint( $request['product_id'] ) );
 
-		// Status.
+		// TODO Fix/Remove in future API version bump. This controls the wrong thing.
+		// https://github.com/woocommerce/woocommerce/issues/15215
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-product-variations-controller.php]: Need to fix the schema for it.

File: includes/api/class-wc-rest-product-variations-controller.php [hunk 2]
```diff
@@ -248,11 +249,17 @@ protected function prepare_object_for_database( $request, $creating = false ) {
...
 		if ( isset( $request['visible'] ) ) {
 			$variation->set_status( false === $request['visible'] ? 'private' : 'publish' );
 		}
 
+		// Status.
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-product-variations-controller.php]: Adding a new parameters should bump the REST API version too.

File: includes/api/class-wc-rest-product-variations-controller.php [hunk 3]
```diff
@@ -248,11 +249,17 @@ protected function prepare_object_for_database( $request, $creating = false ) {
...
+		if ( isset( $request['status'] ) ) {
+			$variation->set_status( get_post_status_object( $request['status'] ) ? $request['status'] : 'draft' );
```
Comment (claudiosanches) [inline includes/api/class-wc-rest-product-variations-controller.php]: On admin interface we allow only `private` and `publish` using the `Enabled` checkbox, not sure if is necessary allow `draft` and other status here.

**Review by claudiosanches — CHANGES_REQUESTED** [review]

File: includes/api/class-wc-rest-product-variations-controller.php (hunk 2 above)
Comment (justinshreve) [inline includes/api/class-wc-rest-product-variations-controller.php]: I'm not sure I agree.. we never did that on WordPress.com or anything else I have worked on. Adding a field doesn't break backwards compatibility. Changing or removing a field does.

File: includes/api/class-wc-rest-product-variations-controller.php (hunk 3 above)
Comment (justinshreve) [inline includes/api/class-wc-rest-product-variations-controller.php]: I think it would be useful. I have one use-case in mind where this would be valuable (even though core doesn't use it yet). In the interface I am working on, if attributes/variation types change, we invalidate existing variations that don't match the new format. I am thinking I would like to mark these as draft instead of immediately trashing them, so they can easily be recovered.

Separate from that, this would 100% match the root products endpoint too.

File: includes/api/class-wc-rest-product-variations-controller.php (hunk 1 above)
Comment (justinshreve) [inline includes/api/class-wc-rest-product-variations-controller.php]: Done.

**Comment by justinshreve** [comment]
Updated the schema for `visible`. Replied to your other two comments.

File: includes/api/class-wc-rest-product-variations-controller.php (hunk 2 above)
Comment (claudiosanches) [inline includes/api/class-wc-rest-product-variations-controller.php]: This creates a difficult set of dependencies.
Like:

> `status` available since REST API v2 and starting only after WooCommerce 3.1.0

So instead of only check for REST API version you need to check the WooCommerce version too.

Comment (justinshreve) [inline includes/api/class-wc-rest-product-variations-controller.php]: I feel like most clients are going to target a particular WooCommerce version anyway. However -- if we are bumping the version I don't think we should do it in this PR. It should be a separate PR since we are also adding the locations endpoints.

Comment (mikejolley) [inline includes/api/class-wc-rest-product-variations-controller.php]: @justinshreve An update to the WordPress.com API rolls out for all clients at once.

And update to the WooCommerce API means multiple versions can be active at the same time since it's self hosted.

Therefore versions need to be increased in our case if the behavior changes.

**Comment by justinshreve** [comment]
I have updated this branch to merge into `feature/rest-api-v3`. I also went ahead and removed the broken `visible` prop since we are targeting a new API version.
